### ISSUE TITLE
Make it possible to package directly from user build dir

### ISSUE THREAD

**Issue by claasd** [conan-io/conan#852]
Hi,

I would like to have the possibility to be able to package directly from my build directory. 
e.g.
```
conan install ..
conan build .. // or cmake commands
conan direct_package ..
```
Thus making it possible to package without the need of having a conan background build. 
I know that the proposed method is to use test_package, but I have some issues with that:

First, I run into long path problems, even when redirecting the data folder. I still have pathes like `d:\cdata\My.Company.BaseLibrary\0.1.0-unreleased-b512\claas\testing\build\63da998e3642b50bee33f4449826b2d623661505` already using up 115 characters of my 250 I have available on Windows. I know that my version is kind of long, but I have this requirement for nightly builds. 

Second, I package the tests in the binary files based on an option. Using the test_package method, I need to build the binaries twice, resulting in long build times. I would like to build once, and then package different DLLs based on options.

Maybe I am looking at this the wrong way. If so, please point me in the correct direction. 

Thanks, Claas

**Comment by memsharded** [comment]
Hi!

I think the command ``conan package`` would do it for you, now it can run locally in user space. The constraint is that it will work on a different folder than the build folder. Please have a look to the reference: http://docs.conan.io/en/latest/reference/commands.html#conan-package

About the long path problems, have you tried the ``short_paths=True`` attribute in Windows?  Check: http://docs.conan.io/en/latest/reference/conanfile.html#short-paths

Regarding the issue of building the things twice, the thing is that the build system should have an option to build only what you need. That would be the same problem of builds always building Debug **AND** Release artifacts, without having an option to choose. If you have a build system option, you will take a conan package option and inject it into the build system at ``build()`` method. If the build system doesn't have such an option, then yes, it will build it twice by default. The correct approach would be to fix the build system. There could be a workaround though: you might use a package to build once, then use two different meta-packages that will do an ``imports`` of the base package, and only import the relevant artifacts.

**Comment by memsharded** [comment]
Hi Claas,

Did you manage to use the ``conan package`` in your user space?
Also, could you manage to avoid re-building the same thing twice for creating two different packages?

Please tell me if you need further help with this issue.

**Comment by claasd** [comment]
Hi,

I still have not a solution that is optimal.
I can use the package command form a different dir, that works.Then the packaged contents are in that folder.  However, I need to upload that package to my conan server. I have not yet figured out how to do that. 
I am not sure yet if my approach is a good way to do that. 

What I am actually trying to to is the following: 
I have a project containing several sub-projects , about half of them are unit-test projects. I would like to build the whole project, including the tests and then package the libraries and the tests separately. 

So basically I am looking for an option that I don't inject in the "build" command, but in the "package" command, without the need of rebuilding (Thus packaging twice). 

I am still not sure how to proceed with this,.

Best,
Claas

**Comment by lasote** [comment]
Maybe you can try to package the whole project and then implement N recipes depending on the big project that only import the needed artifacts and package them. Then you only need to upload the subproject packages. The dependency to the bug project could be declared as private (something like only build dependency) so it won't retrieve the big package when a subproject package is available. Could it work for you at least theoretically?

I haven't tried this setup yet so @memsharded feel free to point some weak points. Maybe we should try it and add it to the conan docs as a `How to` because it's not the first time that this use case appears.

**Comment by memsharded** [comment]
An important question, do you want to package them separately, like, having the same recipe, lets say ``Pkg/0.1@user/channel`` and depending on one package option like ``Pkg:tests``, the resulting binary would be different? I don't understand that setup, because the binary tests will not be able to link (e.g. if the lib was a shared library) with the library for running, because you can't make a package depend on itself with different options.

If you are thinking about having different packages, like ``PkgLib/0.1@user/channel``, and ``PkgTests/0.1@user/channel``, that might be doable. They could both depend on a common ``Pkg/0.1@user/channel`` that did the full build, then ``imports`` the ``Pkg`` desired artifacts, copy them, and lately package them. The dependency to base ``Pkg`` could be private, to avoid retrieving the common package when not necessary. But I still don't get how the ``PkgTests`` would be used. If it required to link with ``PkgLib`` to execute them, then such dependency has to be quite artificially introduced. 

The use case for having the tests in a separate package would be good to know, and your thoughts about the above. Thanks for your feedback!

**Comment by claasd** [comment]
My idea is to have one receipe with one option like `Pkg:withTests`. 
The build method would not use the option, but the package method would.
Example:
```python 
def package(self):
   self.copy("*Pkg.dll", "bin", "bin", keep_path=False)
   if self.options.withTests:
      self.copy("*PkgTest.dll", "bin", "bin", keep_path = False)
```
Thus the option does not control the build, only the packaging.
I can use the `test_package` command twice, once with and once without the options. Thus I would twice build the same stuff. I could add the config to the build, but still the majority of the code would be build twice in the same manner.

**Comment by claasd** [comment]
Probably I need to make two packages out of this, one containing the functionality, one containing the tests, the tests requiring the functionality. 

This however does not reflect the development use case, where I have one CMakeLists.txt with the option BUILD_TESTS. 

It looks like I need to build everything twice, once with and once without tests.

Best,
Claas

**Comment by memsharded** [comment]
The problem with your approach ``Pkg:withTests`` is not that is being built twice, imagine that we could workaround that issue. The real problem is which setup would you use to execute such tests. It is impossible, as they will require the shared library Pkg.dll that lives in another binary of the same recipe, there is no way to do that. A package cannot depend on itself with different options.

I am still interested in the underlying reasons (if possible to know :) ), of doing that. Why do you want to create a binary package with the tests? Most people just create them, run them to ensure the package works as expected, then just package only the library. What is your final setup for running such tests?

I think that I would just add a BUILD_TESTS_ONLY option to the CMakeLists.txt, that will skip building the library and just linking with the precompiled binary provided by the other package (using the cmake conan generator). 

On the other hand, maybe the pragmatic approach can be considered, especially if building in CI, when build time is not that critical, and parallel builds might be possible, building twice is not a big issue.

**Comment by claasd** [comment]
We are required to archive the tests with the package. However, for deploying we need the package without the tests.
Maybe I explained it wrong, I don't require the tests to be a a single package. What I need is to get the package with tests and without tests.
As one of the builds (without) is fully part of the other (with tests), I would like to save Jenkins time end resources.

Best,
Claas

**Comment by memsharded** [comment]
Ok, I think that I understand now. It is necessary to break the 1-1 relation between build-package folder, but I think I might have found a way to do it elegantly. I'll try to propose something on code, to see how it looks like.

**Comment by claasd** [comment]
Sounds good, Thanks!

**Comment by memsharded** [comment]
I have a preliminary working proof of concept, but needs some polishing: https://github.com/memsharded/conan/tree/feature/build_id_info

I don't think we will be able to have it on time for next 0.19 (things are getting wild, 12 PR open atm :) :) ), I am proposing it for next 0.20.

**Comment by claasd** [comment]
Hi,

looks good and very promising :+1.
I just don't get why it's called build_id ;-)

Thanks!
Claas

**Comment by memsharded** [comment]
Well, every package binary has a "package_id", the hash, that matches the "build_id". That is the folder layout for packages is:
```
- export (the copied files at "conan export")
- source (the place where source() is called)
- build/sha_id1 (for certain settings & options)
          /sha_id2 (for other settings & options)
- package/sha_id1
               /sha_id2
```
The ``sha_id`` are the binary packages IDs, so far the relation build-package is one-to-one. With the new method we would be creating a new ID for build, in which many package_ids would map to the same build_id. The logic of that mapping is defined in a ``build_id()`` method. Suggestions for better names welcome :)

**Comment by lasote** [comment]
``build_id`` is not a bad name, but the current ``conan_info()`` name maybe should be ``package_id`` then.

**Comment by memsharded** [comment]
Yes, totally agree. In fact I think that we should rename it, specially if we are documenting it. ``pcakge_id`` has more clear meaning and intent.

**Comment by memsharded** [comment]
Hi @claasd ,

This feature is available in a release candidate:

```bash
$ pip install conan==0.20.0rc1
```
Testing is very appreciated. You can read about some things here: https://github.com/conan-io/docs/blob/develop/howtos/migrate_to_new_environment_management.rst

But the main idea, would be that of: https://github.com/conan-io/conan/pull/964/files#diff-198232cba6994b961f6e5cbc3061bfe1R49

**Comment by claasd** [comment]
Hi,

i can see from the examples how to build the same way for different settings, but I am struggling to get it to work for different options.

Any hints?

**Comment by memsharded** [comment]
Yes, I think you should have an option, like "package_tests=True/False", then:

```python
def build_id(self):
         self.info_build.options.package_tests = "Any"
```
That will generate always the same build_id for any value of the option. You build with one option, the second time you call it with the other value for that option, it should compute the same build_id, then use the same build, and just generate a different package.

**Comment by claasd** [comment]
Hi, 
at a first glance it works as expected. Thanks!
I will do some more testing on Monday.

**Comment by memsharded** [comment]
Great, thanks for reporting. I think we might try to delay the release to Tuesday if necessary, if you detect something on Monday, please report asap. Cheers
